src/style.js

```diff
@@ -297,43 +297,62 @@ export function* groupIndex(I, position, mark, channels) {
   }
 }
 
-// TODO avoid creating a new clip-path each time?
+// TODO Accept other types of clips (paths, urls, x, y, other marks…)?
+// https://github.com/observablehq/plot/issues/181
+export function maybeClip(clip) {
+  if (clip === true) clip = "frame";
+  else if (clip === false) clip = null;
+  else if (clip != null) clip = keyword(clip, "clip", ["frame", "sphere"]);
+  return clip;
+}
+
+function clipDefs({ownerSVGElement}) {
+  const svg = select(ownerSVGElement);
+  const defs = svg.select("defs.clip");
+  return defs.size() ? defs : svg.insert("defs", ":first-child").attr("class", "clip");
+}
+
 // Note: may mutate selection.node!
 function applyClip(selection, mark, dimensions, context) {
   let clipUrl;
   const {clip = context.clip} = mark;
   switch (clip) {
     case "frame": {
-      const {width, height, marginLeft, marginRight, marginTop, marginBottom} = dimensions;
-      const id = getClipId();
-      clipUrl = `url(#${id})`;
-      selection = create("svg:g", context)
-        .call((g) =>
-          g
-            .append("svg:clipPath")
-            .attr("id", id)
-            .append("rect")
-            .attr("x", marginLeft)
-            .attr("y", marginTop)
-            .attr("width", width - marginRight - marginLeft)
-            .attr("height", height - marginTop - marginBottom)
-        )
-        .each(function () {
-          this.appendChild(selection.node());
-          selection.node = () => this; // Note: mutation!
-        });
+      const clips = context.clips ?? (context.clips = new Map());
+      if (!clips.has("frame")) {
+        const {width, height, marginLeft, marginRight, marginTop, marginBottom} = dimensions;
+        const id = getClipId();
+        clips.set("frame", id);
+        clipDefs(context)
+          .append("clipPath")
+          .attr("id", id)
+          .append("rect")
+          .attr("x", marginLeft)
+          .attr("y", marginTop)
+          .attr("width", width - marginRight - marginLeft)
+          .attr("height", height - marginTop - marginBottom);
+      }
+      selection = create("svg:g", context).each(function () {
+        this.appendChild(selection.node());
+        selection.node = () => this; // Note: mutation!
+      });
+      clipUrl = `url(#${clips.get("frame")})`;
       break;
     }
     case "sphere": {
+      const clips = context.clips ?? (context.clips = new Map());
```

Comment:
These probably shouldn’t be exposed publicly on the context. Maybe use a WeakMap keyed by the context instead of assigning to a new property on the context? (Also note that the `Context` interface is publicly documented in `context.d.ts`.)

Reply:
Are there only at most two keys in the clip cache (`frame` and` projection`)?

Reply:
Currently yes, these are the only keys; related: https://github.com/observablehq/plot/issues/1109